ffi/src/lib.rs

```diff
@@ -175,13 +185,43 @@ mod private {
     /// memory, but must only free it by calling [super::drop_bool_slice]. Since the global
     /// allocator is threadsafe, it doesn't matter which engine thread invokes that method.
     unsafe impl Send for KernelBoolSlice {}
+    unsafe impl Send for KernelRowIndexArray {}
 
     /// # Safety
     ///
     /// If engine chooses to leverage concurrency, engine is responsible to prevent data races.
     unsafe impl Sync for KernelBoolSlice {}
+    unsafe impl Sync for KernelRowIndexArray {}
+
+    impl KernelRowIndexArray {
+        /// Converts this slice back into a `Vec<u64>`.
+        ///
+        /// # Safety
+        ///
+        /// The slice must have been originally created `From<Vec<u64>>`, and must not have been
+        /// already been consumed by a previous call to this method.
+        pub unsafe fn into_vec(self) -> Vec<u64> {
+            if self.ptr.is_null() {
+                Default::default()
+            } else {
+                Vec::from_raw_parts(self.ptr, self.len, self.len)
+            }
+        }
+    }
+
+    impl From<DeletionVector> for KernelRowIndexArray {
+        fn from(value: DeletionVector) -> Self {
+            let mut vec = value.row_indexes();
+            vec.shrink_to_fit();
+            let len = vec.len();
+            let boxed = vec.into_boxed_slice();
+            let ptr = Box::into_raw(boxed).cast();
```

Comment:
The cast seems unnecessary (and potentially dangerous)? Doesn't `into_raw` already return the correct type?

Reply:
Yes, otherwise it's a `*mut [u64]` instead of a `*mut u64`

Reply:
What is the difference between `*mut [u64]` and `*mut u64`? Is the former equivalent to C `uint64_t**`? Or merely a way to tell rust that this `unit64_t*` points to an array rather than a single value? If the latter, shouldn't we be keeping the correct type to avoid UB that results from doing array indexing on a ref?

Reply:
I'll be honest, I'm not 100% sure of which it is, I copied from the code above, but maybe we should consider changing it? My expectation is this would be telling C it's a pointer to an array rather than a single value

Reply:
I tried reading up on `*mut [T]` vs. `*mut T` and bogged down at https://internals.rust-lang.org/t/ergonomics-of-raw-pointer-slices/17355/4; my main (but fuzzy) takeaway is that they're _not_ the same thing and we should be careful what we do here. But it also seems like rust already has API calls that build a slice from a pointer+length pair, which suggests that it's only unsafe (rather than UB), as long as the pointer actually points to a sufficiently large slice of memory?

Meanwhile, I _think_ this does the trick without any risky casts:
```rust
let boxed = vec.into_boxed_slice();
let ptr = Box::leak(boxed).as_mut_ptr();
```

Reply:
Done, I updated bool slice to also have this change.